src/classes/RD2_ScheduleService.cls

```diff
@@ -81,6 +84,142 @@ public without sharing class RD2_ScheduleService {
+
+        Date nextDate = schedule.StartDate__c;
+
+        if (schedule.DayOfMonth__c != null && (
+            schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY ||
+            schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_FIRST_AND_FIFTEENTH))
+        {
+            nextDate = Date.newInstance(
+                nextDate.year(),
+                nextDate.month(),
+                schedule.DayOfMonth__c == RD2_Constants.DAY_OF_MONTH_LAST_DAY ?
+                    Date.daysInMonth(nextDate.year(), nextDate.month()) :
+                    Integer.valueOf(schedule.DayOfMonth__c)
+            );
+            nextDate = nextDate >= schedule.StartDate__c ? nextDate : nextDate.addMonths(1);
+        }
+
+        return nextDate;
+    }
+
+    /***
+    * @description Calculates the next donation date >= greaterOrEqualDate
+    * @param schedule Recurring Donation Schedule record
+    * @param greaterOrEqualDate Earliest date limiter (normally today)
+    * @return Date
+    */
+    private Date getNextDonationDateGreaterOrEqualDate(RecurringDonationSchedule__c schedule, Date greaterOrEqualDate) {
+
+        Integer frequency = Integer.valueOf(schedule.InstallmentFrequency__c);
+        Date firstValid = getFirstValidDonationDate(schedule);
+
+        if (firstValid >= greaterOrEqualDate) {
+            return firstValid;
+        }
+
+        Integer unitsBetween;
+        Integer remains;
+        Date adjusted;
+
+        if (schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY || schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_FIRST_AND_FIFTEENTH) {
+
+            unitsBetween = firstValid.monthsBetween(greaterOrEqualDate);
+            remains = Math.mod(unitsBetween, frequency);
+            adjusted = firstValid.addMonths(unitsBetween);
+
+            if (remains > 0) {
+                adjusted = adjusted.addMonths(1);
+            }
+
+            if (adjusted < greaterOrEqualDate) {
+                adjusted = adjusted.addMonths(frequency);
+            }
+
+            if (schedule.DayOfMonth__c == RD2_Constants.DAY_OF_MONTH_LAST_DAY) {
+                adjusted = Date.newInstance(adjusted.year(), adjusted.month(), Date.daysInMonth(adjusted.year(),adjusted.month()));
+            }
+        }
+
+        else if (schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_YEARLY) {
+
+            unitsBetween = firstValid.monthsBetween(greaterOrEqualDate);
+            remains = Math.mod(unitsBetween, frequency*12);
+            adjusted = firstValid.addYears(unitsBetween / (12));
+
+            if (remains > 0) {
+                adjusted = adjusted.addYears(frequency);
+            }
+
+            if (adjusted < greaterOrEqualDate) {
+                adjusted = adjusted.addYears(frequency);
+            }
+        }
+
+        else if (schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_WEEKLY) {
```

Comment:
Would it be easy to add another block here to support `INSTALLMENT_PERIOD_DAILY`? It's something that can exist if customers defined that as a custom schedule in RD1.